synapse/crypto/keyring.py

```diff
@@ -325,16 +354,20 @@ def _attempt_key_fetches_with_fetcher(self, fetcher, remaining_requests):
             remaining_requests (set[VerifyKeyRequest]): outstanding key requests.
                 Any successfully-completed requests will be reomved from the list.
         """
-        # dict[str, set(str)]: keys to fetch for each server
-        missing_keys = {}
+        # dict[str, dict[str, int]]: keys to fetch.
+        # server_name -> key_id -> min_valid_ts
+        missing_keys = defaultdict(dict)
+
         for verify_request in remaining_requests:
             # any completed requests should already have been removed
             assert not verify_request.deferred.called
-            missing_keys.setdefault(verify_request.server_name, set()).update(
-                verify_request.key_ids
-            )
+            keys_for_server = missing_keys[verify_request.server_name]
+            for key_id in verify_request.key_ids:
+                current_min_ts = keys_for_server.get(key_id, -1)
+                if current_min_ts < verify_request.minimum_valid_until_ts:
+                    keys_for_server[key_id] = verify_request.minimum_valid_until_ts
```

Comment:
Does the remote server respond with the key with the latest `valid_until_ts` even if that is less than the requested if it can't find a later one? If not will that cause problems where the key may have been valid for some of the key requests?

Reply:
> Does the remote server respond with the key with the latest valid_until_ts even if that is less than the requested if it can't find a later one? If not will that cause problems where the key may have been valid for some of the key requests?

Hum, apparently a notary server will not respond with such a key. And yes, it probably will. I'll try and get that changed on the notary server impl before landing this.

Reply:
> Can haz comment pls?

done

> Maybe this can be written as:

I'm not entirely convinced it's clearer, but have tweaked it anyway.

Reply:
> Hum, apparently a notary server will not respond with such a key. 

This is apparently incorrect, as tested by https://github.com/matrix-org/sytest/pull/620.